### ISSUE TITLE
Replace lximage-qt screenshot functionality.

### ISSUE THREAD

**Issue by jleclanche** [lxqt/screengrab#42]
This has been on my mind for a long time. The screenshot functionality in lximage-qt makes very little sense and screengrab works quite well (although the code quality is sub-par for LXQt, but that's fixable).

High level thoughts?

**Comment by agaida** [comment]
Full ACK for moving it - and full ACK for replacing the screenshot functionality in lximage-qt after a code cleanup.

**Comment by agaida** [comment]
if nothing speaks against i would like to move it now

**Comment by jleclanche** [comment]
Go for it.

**Comment by luis-pereira** [comment]
Do it.

**Comment by agaida** [comment]
just moved

**Comment by agaida** [comment]
Retitled

**Comment by wxl** [comment]
Consistent with this, can we please drop lximage-qt-screenshot.desktop?

**Comment by agaida** [comment]
Maybe it is worth to implement a direct upload mechanism - right now "enable external viewer", double click and upload with lximage-qt works fine, but is a bit counterintuitive. 

Second thought - we should drop the lximage-qt screenshot functionality and use screengrap or any other working tool directly from lximage-qt if possible (I'm happy to say that i have no plan how to implement, but i hate duplicated code and functionality)

just an addition for the ancient bug

**Comment by tsujan** [comment]
> Second thought - we should drop the lximage-qt screenshot functionality and use screengrap

A year ago, I might disagree but now, I agree.

However, there is a problem: many may use lximage-qt for taking screenshots. If we remove that functionality, they'll hate us. We can't tell them to install this or that app instead. I can put myself in their shoes, using swear words...

**Comment by agaida** [comment]
hmm - no, nobody will use swearwords if we do it right

we can call screengrab from lximage-qt, if installed - and also transfer the screenshot to lximage-qt in the same routine -- so nothing changes much

**Comment by tsujan** [comment]
> we can call screengrab from lximage-qt, if installed

"If installed"; that's the problem.

As a developer, I like a cleaner code; so, I agree. As a user, I'd say, "Why did they remove a working code? Gnome mentality?" So, I think we could still wait. The screenshot code is functional, isolated and harmless.

BTW, we shouldn't add more "editing codes" either. The current image editing code has bad bugs. I prefer a good image viewer to a buggy hybrid.

**Comment by agaida** [comment]
yes, it should be fixed or removed - btw a good choice for screenshots with basic editing done nearly right is 'flameshot' - i like the interface to add text, some arrows, blur sensible things etc - the only thing that they don't have implemented yet is the delayed screenshot functionality (try to make a shot of an open menu ...)

RE: if installed - it's like in desktop files :
```
[Desktop Entry]
Version=1.0
Name=Workspaces
Comment=Configure layout, names and margins
....
TryExec=xfwm4-workspace-settings
Exec=xfwm4-workspace-settings
Icon=xfce4-workspaces
Terminal=false
Type=Application
Categories=Settings;DesktopSettings;System;
StartupNotify=true
OnlyShowIn=LXQt;
```
So if xfwm4 isn't installed - my desktop file will not be shown in the menu and our controlcenter

**Comment by tsujan** [comment]
> yes, it should be fixed or removed 

I prefer to fix it without adding to it. I need only one thing: time.

> So it would be only a question of packaging - lximage-qt should recommend...

It seems like a good idea. But please first do it and, only after some time, remove the screenshot code.

**Comment by agaida** [comment]
for sure - the packaging part is easy, the implementation possible not

**Comment by Filip98** [comment]
I just realised screengrab exists... why not just make it an optional dependency of lximage-qt, people who read won't complain

**Comment by tsujan** [comment]
Very sorry that I have to change my mind again -- but based on facts:

Once in a while and for the fun of it, I use Enlightenment, especially when a new release comes. It's also useful for testing libfm-qt based apps outside a Qt or GTK environment.

Enlightenment has a minimal screenshot functionality. scrot can't include E title-bar and screengrab can't even take a window screenshot (it takes the whole screen in the window mode too). Only lximage-qt works fine under E.

I don't think screengrab can be "fixed" (although I'll look into it). But lximage-qt's screenshot functionality is completely based on X11 and that may be the reason for its success in this case.

**Comment by agaida** [comment]
nobody want to fix screengrab to be honest - really - I had this idea years ago and asked politely if it would be possible to remove the screenshot functionality from lximage-qt without any compensation - we should just follow the ancient standard and provide tools which do one thing perfectly - screengrab is nearly perfect in doing screenshots - so there is nothing to change I guess. What we should change however is the integration into lximage-qt. 

And of course the integrational part of uploading things has to be separated from any sceenshot utility - that is what I preached for years - my suggestion is to implement the screenschot things in lximage-qt like a simple call, the second suggestion is to implement the upload functionality outside and indipendent from any tool - it could be very useful.

**Comment by jleclanche** [comment]
@agaida Spectacle works fantastically well. It's probably a better bet to fork spectacle to try to reduce its dependencies on kde stuff.

**Comment by tsujan** [comment]
Screengrab works like a charm; I don't think we need Spectacle, which can't even be styled outside KDE (it only uses Fusion) and is QML-based. Most probably, the issue I mentioned above is caused by peculiarities of E and is reproducible with Spectacle too.

**Comment by tsujan** [comment]
> screengrab is nearly perfect in doing screenshots - so there is nothing to change I guess.

@agaida No app can be perfect ;) I like Screengrab. But like any other app, it needs improvements. Taking screenshots under Wayland is just one example.

My comment about E may have been confusing. I try to correct that here.

IMHO, Screengrab is nice as LXQt's screenshot utility and its development can be continued like other LXQt components. I wouldn't remove the screenshot code of lximage-qt soon (and I think no one would) but only after a sufficiently long time. As for uploading images, I don't know; have never found time to take a look at Screengrab's uploading code and, apparently, others haven't either.

**Comment by tsujan** [comment]
> nobody want to fix screengrab to be honest 

I didn't understand that. Shouldn't we keep developing it?

After making 7 patches for screengrab in 2018, I didn't find time for it but didn't consider it abandoned either.

**Comment by wxl** [comment]
> Once in a while and for the fun of it, I use Enlightenment

I like masochism as much as the next guy, @tsujan, but this is really taking it too far.

> Most probably, the issue I mentioned above is caused by peculiarities of E

All seriousness aside, I actually do like Englightment a lot, but it does have its peculiarities. I don't think we should make decisions in general based on compatibility with E.

>>    nobody want to fix screengrab to be honest

> I didn't understand that. Shouldn't we keep developing it?

I think @agaida was expressing his frustration with the upload module, which has been a [known problem since 2016](https://github.com/lxqt/screengrab/issues/28) and ultimately led to [entire removal of the upload module](https://github.com/lxqt/screengrab/issues/179) by the man himself. He may very well be saying that *he* doesn't want to fix it :stuck_out_tongue_winking_eye:

**Comment by tsujan** [comment]
> but this is really taking it too far.

@wxl You're right. I lost perspective.

> I don't think we should make decisions in general based on compatibility with E.

I found and fixed important bugs in pcmanfm-qt/libfm-qt only under E -- they were hidden with LXQt. E's peculiarities can be illuminating. But, every time I start to use E, an intolerable E bug shows up and I return to the comfort of LXQt+KWin after reporting it. This time, I have to remain there a bit longer (for the report).

There's an interesting similarity between E and LXQt: fairly good development in spite of lack of manpower. The difference: I don't know of an intolerable bug in LXQt ;)

**Comment by tsujan** [comment]
For the record:

I found the cause of Screengrab's behavior under E: the conditions for taking a window screenshot are too strong. I loosened one condition just a little (without removing it) and it worked.

I tried Spectacle too: contrary to what I guessed at https://github.com/lxqt/screengrab/issues/42#issuecomment-565800118, its behavior was like lximage-qt. So, I think the condition in question should be loosened. I'll do so after https://github.com/lxqt/screengrab/pull/190 is decided.

Again, E was informative :)

**Comment by agaida** [comment]
RE: Nobody want to fix screengrab - wrong context - don't fix things without reasons. Removing qxl and uploader code was long overdue - and any improvement is welcome. What i really want is 
a) a screengrab integration into lximage-qt
b) an indipendent uploading thing that is easy to configure and might be used with other LXQt components or stand alone if possible.

**Comment by tsujan** [comment]
Not relevant anymore.

We kept lximage-qt's screenshot functionality for X11 and improved screengrab for both X11 and Wayland.
